src/context.ts

```diff
@@ -64,11 +75,44 @@ const MessageSubTypes = [
   'poll',
   'forward_date',
 ] as const
+export type MessageProps = {
+  [key in MessageSubTypesUnionMapped]: {
+    message: {
+      [k in MapSubTypeBack<key>]: UnionToIntersection<tt.Message>[k]
+    }
+  }
+}
```

Comment:
Don't bother reviewing this further. First, I must work around the problem of having union types that are too large to be represented. It is possible to manually pre-compute some types in some critical inference steps. This cuts down the compile time significantly, too.

The above type is one of the magical parts that I abandoned. It is correct but too slow. I can still explain it, though, it is crucial that I'm not the only one who knows what's going on :wink:

The general idea is straightforward: instead of having middleware for `TContext extends Context`, we intersect the context subclass with a type that requires the properties that we assume to be present. Here is a concrete example of this.

```ts
bot.on('message', mw)
```

We can now infer that `mw` has to be middleware of type `Middleware<Context & { message: tt.Message }>`. This makes sure that the `message` property is given and has the correct type. (Note that `Middleware<Context & { message: object }>` would also be enough because the intersection of `Context['message'] & object` simply boils down to `Context['message']`.)

Thinking this through, we therefore have to map the `'message'` in the first argument of `on` to this type. As we always intersect with `TContext extends Context` independent of `'message'`, we only need to obtain the type `{ message: tt.Message }` from `'message'`.

There are three types alike to the one you quoted, one of them does exactly this.

Regarding subtypes of messages, we simply need to take everything one step further and also intersect with the message property in order to assure the presence of the corresponding properties during type inference. The type you quoted does exactly that: it takes a `tt.Message` and picks out the respective property, hence creating an object that requires it. This resulting object can be intersected with `TContext extends Context` such that we obtain the correct type for `mw`.

The mapping of types with `MessageSubTypesUnionMapped` and `MapSubTypeBack` solves that problem that some message subtypes have different names in the event name than in the property name. This mapping is specified in `MessageSubTypesMapping`, i.e. `forward` is used to register middleware for message forwards, but `forward_date` is actually the property that we have to require on `TContext extends Context`. We derive a reverse mapping of the `MessageSubTypesMapping` by inventing a type that swaps keys and values in an object. This can then be used during type inference.

As a said, the PR works fine and I'm halfway through annotating the method signatures, but these comprehensive union types are simply a too large thing to rely on in such advanced use cases. I'll ping you once I have something that is both correct _and_ fast enough to the pleasant to use.